pkg/controller/utils/capability.go

```diff
@@ -0,0 +1,286 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+	"github.com/oam-dev/kubevela/apis/types"
+	mycue "github.com/oam-dev/kubevela/pkg/cue"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+	"github.com/oam-dev/kubevela/pkg/utils/common"
+)
+
+const (
+	// UsageTag is usage comment annotation
+	UsageTag = "+usage="
+	// ShortTag is the short alias annotation
+	ShortTag = "+short"
+)
+
+// CapabilityDefinitionInterface is the interface for Capability (WorkloadDefinition and TraitDefinition)
+type CapabilityDefinitionInterface interface {
```

Comment:
I think three implementation structs are a bit duplicate, can we merge them?( like:  https://github.com/oam-dev/kubevela/blob/a8337921bc13d9efee8fbfa28c27af5045203f95/pkg/oam/util/template.go#L40)

Reply:
Okay, let's merge them after your PR is merged.

Reply:
As @WangYikewyk said , the code can be more clean here